x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/TopValuesListLongAggregator.java

```diff
@@ -0,0 +1,175 @@
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.LongBlock;
+import org.elasticsearch.compute.operator.DriverContext;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.search.sort.RawBucketedSort;
+import org.elasticsearch.search.sort.SortOrder;
+
+/**
+ * Aggregates the top N field values for long.
+ */
+@Aggregator({ @IntermediateState(name = "topValuesList", type = "LONG_BLOCK") })
+@GroupingAggregator
+class TopValuesListLongAggregator {
+    public static SingleState initSingle(BigArrays bigArrays, int limit, boolean ascending) {
+        return new SingleState(bigArrays, limit, ascending);
+    }
+
+    public static void combine(SingleState state, long v) {
+        state.add(v);
+    }
+
+    public static void combineIntermediate(SingleState state, LongBlock values) {
+        int start = values.getFirstValueIndex(0);
+        int end = start + values.getValueCount(0);
+        for (int i = start; i < end; i++) {
+            combine(state, values.getLong(i));
+        }
+    }
+
+    public static Block evaluateFinal(SingleState state, DriverContext driverContext) {
+        return state.toBlock(driverContext.blockFactory());
+    }
+
+    public static GroupingState initGrouping(BigArrays bigArrays, int limit, boolean ascending) {
+        return new GroupingState(bigArrays, limit, ascending);
+    }
+
+    public static void combine(GroupingState state, int groupId, long v) {
+        state.add(groupId, v);
+    }
+
+    public static void combineIntermediate(GroupingState state, int groupId, LongBlock values, int valuesPosition) {
+        int start = values.getFirstValueIndex(valuesPosition);
+        int end = start + values.getValueCount(valuesPosition);
+        for (int i = start; i < end; i++) {
+            combine(state, groupId, values.getLong(i));
+        }
+    }
+
+    public static void combineStates(GroupingState current, int groupId, GroupingState state, int statePosition) {
+        current.merge(groupId, state, statePosition);
+    }
+
+    public static Block evaluateFinal(GroupingState state, IntVector selected, DriverContext driverContext) {
+        return state.toBlock(driverContext.blockFactory(), selected);
+    }
+
+    public static class GroupingState implements Releasable {
+        private final RawBucketedSort.ForLongs sort;
+
+        private GroupingState(BigArrays bigArrays, int limit, boolean ascending) {
+            this.sort = new RawBucketedSort.ForLongs(bigArrays, ascending ? SortOrder.ASC : SortOrder.DESC, limit);
+        }
+
+        public void add(int groupId, long value) {
+            sort.collect(groupId, value);
+        }
+
+        public void merge(int groupId, GroupingState other, int otherGroupId) {
+            // TODO: Make a merge method in RawBucketedSort?
+            other.sort.getValues(otherGroupId).forEach(value -> sort.collect(groupId, value));
```

Comment:
:+1:

Reply:
It should be very very efficient to merge actually. It'll amount to two min-heaps.